### DISCUSSION TITLE
Typescript Circular properties causing typed Any

### DISCUSSION THREAD

**Discussion by sethdumaguin** [mobxjs/mobx-state-tree#1879] [General]
Hi! 

I am trying to develop some type of Folder structures where I defined these three entities:

`File`: the files in a folder,
`FolderItem`: the items that can be in a folder, which is a union of `File` and `Folder` since a folder can have both.
`Folder`: a folder that holds `FolderItem`.

The problem that I am seeing is the circular dependency here is causing `Folder` and `FolderItem` to be typed as any. I was wondering if there's a way to go around this. Here is a codesandbox link: https://codesandbox.io/s/little-monad-oim34o?file=/src/Folder.ts.

I appreciate the help!

**Comment by EmilTholin** [comment]
Hi @sethdumaguin!

I don't see an obvious solution in your example, but [this part of the documentation](https://mobx-state-tree.js.org/tips/circular-deps) and [this neat trick by kresli](https://github.com/mobxjs/mobx-state-tree/issues/943#issuecomment-442603532) might give some inspiration.

```ts
import { IAnyModelType, Instance, types } from "mobx-state-tree";

const Node = types
  .model({
    x: 5,
    me: types.maybe(types.late((): IAnyModelType => Node))
  })
  .views((self) => ({
    getMe() {
      return self.me as Instance<typeof self> | undefined;
    }
  }));

const node = Node.create({
  x: 5,
  me: {
    x: 4
  }
});
const me = node.getMe();

// Works!
me?.x
// Property 'y' does not exist on type...
me?.y
```
